Rewrite opt for combining consecutive byte loads

IBMJimmyk opened the pull request:
Changes were made inside the sequentialStoreSimplification. Despite the name
of the optimization, the changes are targeted at sequential loads.

Rewrote the optimization that identifies byte loads from conseuctive memory
addresses and replaces them with a wider load such as an iload or lload. The
new version greatly expands the number of tree patterns indentified by the
optimization. The new wide load optimization works with 2-8 consecutive byte
loads and in both big and little endian order.

The original version of the opt is still available under the env var
TR_UseOldSeqLoadOpt.

Moved sequentialStoreSimplificationGroup later in the optimization order so
that the trees are more likely to be in the right shape for the optimization.

Moved sequentialStoreSimplification out of Optimizations.enum to OMR where it
is used for the tracing option.

Closes: #11250
Signed-off-by: jimmyk <jimmyk@ca.ibm.com>

IBMJimmyk commented:
This PR needs to go in at the same time as this OMR PR since code was moved from OpenJ9 to OMR:
https://github.com/eclipse/openj9/pull/11251

@gita-omr Could you take a look at these changes?

IBMJimmyk commented:
`sequentialStoreSimplification` is no longer being moved out of `Optimizations.enum` to OMR. This should mean this change can go in independently of the OMR changes.

gita-omr commented:
Very nice code. I really like how general it is. I have some minor comments/questions but I will post them a bit later (busy with defects this week).

gita-omr commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+    *   bloadi
+    *     aladd
+    *       aload/aloadi
+    *       lconst        <<--- displacementNode, return value
```
Do we allow byte loads without the basePointerNode? How would we match then nodes that have it and the ones that don't?

IBMJimmyk replied:
Yes. This happens if the index into the array are constants like `a[0], a[1], a[2], a[3]`. In this case `basePointerNode` is treated as being NULL since it does not exist. If it is NULL in one of the byte loads that are being combined it must be NULL for the others as well to be a match. If it is NULL for one load and not for another, it is not considered a match.

gita-omr commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+       * multValue is used to determine the significance of the byte load.
+       * If two byte loads have the same multValue, the opt bails out.
+       */
+      for (int j = 0; j < byteCount; j++)
```
Is it possible to replace this loop with mapping from multValue to j and check that j is with the right range and that there no duplicates? Then the following loop would not be needed either

IBMJimmyk replied:
Yes. I can change this.

gita-omr commented:
I think some tests need to be added. Perhaps convert your unit tests?

IBMJimmyk commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+      return currentTree;
+
+   visitedNodes->set(currentNode->getGlobalIndex());
+
```
I can add more comments.

The transformations are all done at the end of `generateArraycopyFromSequentialLoads`. Everything before the comment that says "All checks have passed so perform the transformation." is analysis.

IBMJimmyk commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
-            }
-         }
+      /*
+       * The initial tree structure seems good at this point. generateArraycopyFromSequentialLoads will do some additional
```
That comment just means all checks have passed so far. But, there are still more checks to be done inside `generateArraycopyFromSequentialLoads` so it is not known if it is actually good yet. I can try adding more comments here as well.

IBMJimmyk commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2200 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+            if (!isValidSeqLoadMulOrShl(comp, trace, childrenArray[i]))
+               return false;
+            break;
+         case TR::iand:
```
For the loaded bytes other than the most significant byte, they need to be treated as unsigned bytes or the sign bits may get in the way when they are combined into a larger value. This is done by `and` with `0xff`. I think `treeSimplification` will normally identify the pattern of a byte conversion like `b2i` feeding into an `and` with `0xff` and convert it into a `bu2i` instead. But, if it doesn't, this pattern will still work.

IBMJimmyk commented:
Previous comments related to adding more comments have been addressed. I also change how `multValue` is handled so that a loop is not needed. I also rebased the code onto the latest openj9.

Comment about needed tests still needs to be addressed.

gita-omr commented:
Thanks @IBMJimmyk for addressing the comments. @0xdaryl would you like to take a look at this new common optimization?

IBMJimmyk commented:
I just added a bunch of new tests to exercise the new optimization. I'm not entirely sure I got all the formatting right here so please take a look and tell me if I made any mistakes.

fjeremic commented on test/functional/JIT_Test/src/jit/test/tr/SeqLoadSimplification/SeqLoadSimplificationTest.java:
```diff
@@ -0,0 +1,935 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package jit.test.tr.SeqLoadSimplification;
```
Does this need to be in it's own directory? Do we envision adding more files here? If not we should move it out.

Great set of tests by the way! Readers take note.

IBMJimmyk replied:
Thanks.

I don't expect to add more files here. However, I don't really see a better place to put it. I tried looking at other tests to see how they organize their files but a lot of them are single files in a directory as well.

fjeremic replied:
Can we simply move `SeqLoadSimplificationTest.java` into `test/functional/JIT_Test/src/jit/test/tr/`?

IBMJimmyk replied:
Yes, that's possible. In my latest changed I've now moved it out.

gita-omr commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2233 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+         {
+         /* byteCount is 4, final result is an int, no byteswaps */
+         if (trace) traceMsg(comp, "Recreating rootNode (%p) as iloadi.\n", rootNode);
+         TR::Node::recreateWithSymRef(rootNode, TR::iloadi, newLoadChildNode->getSymbolReference());
```
Will we have an iloadi node with a symbol reference of type byte(Int8) ? I think this can lead to some complications in the optimizer and the codegens.

IBMJimmyk replied:
I'm not sure if this is a problem. I will look into it.

IBMJimmyk replied:
In the latest changes I changed the symref to being a generic int array shadow symref.

gita-omr reviewed (changes requested):
Once again, nice code and this is only a partial review since I need to save my comments.

IBMJimmyk commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2233 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
-            }
-         }
+      /*
+       * The initial tree structure seems good at this point. generateArraycopyFromSequentialLoads will do some additional
```
I decided to just remove this comment since it keeps causing problems. There's already a comment a few lines earlier that mentions what `isValidSeqLoadCombine` and `generateArraycopyFromSequentialLoads` do. Those functions themselves also have large comments that provide the details on what they are doing.

IBMJimmyk commented on runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:
```diff
@@ -677,621 +684,2233 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
+    *     aladd
+    *       aload/aloadi
+    *       lsub/ladd
+    *         **basePointerNode**
```
I copied the name from the original implementation. I can rename it to base offset.

IBMJimmyk commented:
I think I've addressed all the comments so far.

@gita-omr Did you want to look over this again?

@vijaysun-omr Could you look over the Symref changes I made and confirm that I am using the generic int array shadow symref correctly? The changes are all inside `generateArraycopyFromSequentialLoads` in the latest commit.

vijaysun-omr commented:
@IBMJimmyk the usage of generic int shadow looks okay to me. I feel that you may want to do one small sanity test by checking the aliasing sets (traceAliases should allow you to see that) that you get and make sure the generic int shadow created by your optimization does conservatively alias the array shadow sym refs that were in use originally. This would ensure that optimizations do not do the wrong thing with respect to moving stores to the array shadow sym refs past the load of the generic int shadow sym ref or vice versa.

IBMJimmyk commented:
I tried running a small test I wrote with the `traceAliases` option. Under `Symbol References with Aliases`, I see the following for alias info:
```
Symref #248 <array-shadow>
   Use Aliases: NULL
   Usedef Aliases: 00007282BA340568 {248, 255, 265, 407, 409, 415}

Symref #415 <generic int shadow>
   Use Aliases: NULL
   Usedef Aliases: 00007282BA3408C8 {248, 407, 409, 415}
```
`Symref #248 <array-shadow>` is the symref that was on the `bloadi` nodes before the transformation. When the transformation replaces them with a `iloadi`, the `iloadi` is assigned `Symref #415 <generic int shadow>`. From the Usedef alias info it can be seen that old `#248` and the new `#415` do alias each other.

gita-omr reviewed (commented):
Does this shadow work for all integral types, e.g. short, int, long loads/stores? 
@vijaysun-omr

vijaysun-omr commented:
Yes I believe it does but Jimmy can double check by looking at the existing usages.

IBMJimmyk commented:
`generic int shadow` is used even for types other than int. There is no `generic long shadow` or `generic short shadow` to begin with.

gita-omr commented:
> `generic int shadow` is used even for types other than int. There is no `generic long shadow` or `generic short shadow` to begin with.

I thought the size of the shadow is important since some optimizations/evaluators might be checking the size of the symbol reference on a node vs node's opcode.

vijaysun-omr commented:
Generic int shadows are not optimized in all cases where normal field shadows are, e.g. there is no signature etc. on them and so an optimization like VP might not obtain anything from the sym ref as such over and above what it gets from the IL opcode. Other places in the optimizer that look at the size of the shadow as opposed to the node do not come to mind, but I'm unsure if you had something specific in mind when you commented.

Regardless, @IBMJimmyk change did not create this precedent, it was already there in the code for many years. He was just adding another use of the generic int shadow for his feature but there were other places that used this sym ref with a wider IL opcode before his change.

gita-omr commented:
Thanks @vijaysun-omr for the background. I don't have any specific example right now, just recall from previous experience that opcode/symref mismatch *might* cause an issue. Given what you described, I am fine with the use of generic int shadows in this case until we find more specific example where it can cause a problem.

IBMJimmyk commented:
I rebased the changes on the latest master and squashed all the commits together.

genie-openj9 commented:
Can one of the admins verify this patch?

gita-omr commented:
@IBMJimmyk @zl-wang , was there any reason for not merging it?

IBMJimmyk commented:
I was doing some performance testing and then got sidetracked. The testing was finished and no problems came up so this can be merged in if there are no other comments otherwise.

zl-wang commented:
Jenkins test sanity all jdk8

zl-wang commented:
Jenkins test sanity all jdk8

IBMJimmyk commented:
All the checks passes, can this be merged in?
